x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,218 +232,407 @@ public WildcardFieldType clone() {
+                return;
+            }
+            // Break fragment into multiple Ngrams
+            TokenStream tokenizer = WILDCARD_ANALYZER.tokenStream(name(), fragment);
+            CharTermAttribute termAtt = tokenizer.addAttribute(CharTermAttribute.class);
+            // If fragment length < NGRAM_SIZE then it is not emitted by token stream so need
+            // to initialise with the value here
+            String lastUnusedToken = fragment;
+            try {
+                tokenizer.reset();
+                boolean takeThis = true;
+                // minimise number of terms searched - eg for "12345" and 3grams we only need terms
+                // `123` and `345` - no need to search for 234. We take every other ngram.
+                while (tokenizer.incrementToken()) {
+                    String tokenValue = termAtt.toString();
+                    if (takeThis) {
+                        tokens.add(tokenValue);
+                        lastUnusedToken = null;
+                    } else {
+                        lastUnusedToken = tokenValue;
+                    }
+                    // alternate
+                    takeThis = !takeThis;
+                    if (tokens.size() >= MAX_CLAUSES_IN_APPROXIMATION_QUERY) {
+                        lastUnusedToken = null;
+                        break;
+                    }
+                }
+                if (lastUnusedToken != null) {
+                    // given `cake` and 3 grams the loop above would output only `cak` and we need to add trailing
+                    // `ake` to complete the logic.
+                    tokens.add(lastUnusedToken);
+                }
+                tokenizer.end();
+                tokenizer.close();
+            } catch (IOException ioe) {
+                throw new ElasticsearchParseException("Error parsing wildcard regex pattern fragment [" + fragment + "]");
+            }
         }
+        
 
-        private void addClause(String token, BooleanQuery.Builder bqBuilder) {
+        private void addClause(String token, BooleanQuery.Builder bqBuilder, Occur occur) {
             assert token.codePointCount(0, token.length()) <= NGRAM_SIZE;
-            if (token.codePointCount(0, token.length()) == NGRAM_SIZE) {
+            int tokenSize = token.codePointCount(0, token.length());
+            if (tokenSize < 2 || token.equals(WildcardFieldMapper.TOKEN_END_STRING)) {
+                // there's something concrete to be searched but it's too short
+                // Require verification.
+                bqBuilder.add(new BooleanClause(new MatchAllButRequireVerificationQuery(), occur));
+                return;
+            }
+            if (tokenSize == NGRAM_SIZE) {
                 TermQuery tq = new TermQuery(new Term(name(), token));
-                bqBuilder.add(new BooleanClause(tq, Occur.MUST));
+                bqBuilder.add(new BooleanClause(tq, occur));
             } else {
```

Comment:
This shouldn't be possible ? Maybe add an assert that tokenSize should be less than or equals to NGRAM_SIZE at this point ?

Reply:
Not sure I understand - do you mean the else is not possible? It should catch when tokenSize ==2. I think we agreed that `ab*` was worth running but not `a*`

>Maybe add an assert that tokenSize should be less than or equals to NGRAM_SIZE at this point ?

line 515 has this assert?

Reply:
Sorry I missed the too short condition.